File: inventory-service/src/main/java/com/ecommerce/inventory/controller/InventoryController.java
package com.ecommerce.inventory.controller;

import com.ecommerce.inventory.dto.*;
import com.ecommerce.inventory.service.InventoryService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/inventory")
@RequiredArgsConstructor
@Slf4j
@Tag(name = "Inventory Management", description = "APIs for product catalog and inventory management")
public class InventoryController {
    
    private final InventoryService inventoryService;
    
    @PostMapping("/products")
    @Operation(summary = "Create a new product")
    public ResponseEntity<ProductResponse> createProduct(@Valid @RequestBody ProductRequest request) {
        log.info("POST /api/inventory/products - Creating product: {}", request.getName());
        ProductResponse response = inventoryService.createProduct(request);
        return new ResponseEntity<>(response, HttpStatus.CREATED);
    }
    
    @GetMapping("/products/{id}")
    @Operation(summary = "Get product by ID")
    public ResponseEntity<ProductResponse> getProductById(@PathVariable Long id) {
        log.info("GET /api/inventory/products/{} - Fetching product", id);
        ProductResponse response = inventoryService.getProductById(id);
        return ResponseEntity.ok(response);
    }
    
    @GetMapping("/products")
    @Operation(summary = "Get all products")
    public ResponseEntity<List<ProductResponse>> getAllProducts(
            @RequestParam(required = false) String category) {
        log.info("GET /api/inventory/products - Fetching products");
        
        List<ProductResponse> response;
        if (category != null && !category.isEmpty()) {
            response = inventoryService.getProductsByCategory(category);
        } else {
            response = inventoryService.getAllProducts();
        }
        
        return ResponseEntity.ok(response);
    }
    
    @PutMapping("/products/{id}")
    @Operation(summary = "Update product")
    public ResponseEntity<ProductResponse> updateProduct(
            @PathVariable Long id,
            @Valid @RequestBody ProductRequest request) {
        log.info("PUT /api/inventory/products/{} - Updating product", id);
        ProductResponse response = inventoryService.updateProduct(id, request);
        return ResponseEntity.ok(response);
    }
    
    @DeleteMapping("/products/{id}")
    @Operation(summary = "Delete product")
    public ResponseEntity<Map<String, String>> deleteProduct(@PathVariable Long id) {
        log.info("DELETE /api/inventory/products/{} - Deleting product", id);
        inventoryService.deleteProduct(id);
        return ResponseEntity.ok(Map.of("message", "Product deleted successfully"));
    }
    
    // Inter-service communication endpoints
    @PostMapping("/check-stock")
    @Operation(summary = "Check stock availability (for inter-service communication)")
    public ResponseEntity<StockCheckResponse> checkStock(@Valid @RequestBody StockCheckRequest request) {
        log.info("POST /api/inventory/check-stock - Checking stock for product: {}", request.getProductId());
        StockCheckResponse response = inventoryService.checkStock(request);
        return ResponseEntity.ok(response);
    }
    
    @PostMapping("/reserve-stock")
    @Operation(summary = "Reserve stock (for inter-service communication)")
    public ResponseEntity<Map<String, String>> reserveStock(@RequestBody StockCheckRequest request) {
        log.info("POST /api/inventory/reserve-stock - Reserving stock for product: {}", request.getProductId());
        inventoryService.reserveStock(request.getProductId(), request.getQuantity());
        return ResponseEntity.ok(Map.of("message", "Stock reserved successfully"));
    }
    
    @PostMapping("/release-stock")
    @Operation(summary = "Release reserved stock (for inter-service communication)")
    public ResponseEntity<Map<String, String>> releaseStock(@RequestBody StockCheckRequest request) {
        log.info("POST /api/inventory/release-stock - Releasing stock for product: {}", request.getProductId());
        inventoryService.releaseStock(request.getProductId(), request.getQuantity());
        return ResponseEntity.ok(Map.of("message", "Stock released successfully"));
    }
    
    @PostMapping("/confirm-stock")
    @Operation(summary = "Confirm stock purchase (for inter-service communication)")
    public ResponseEntity<Map<String, String>> confirmStock(@RequestBody StockCheckRequest request) {
        log.info("POST /api/inventory/confirm-stock - Confirming stock for product: {}", request.getProductId());
        inventoryService.confirmStock(request.getProductId(), request.getQuantity());
        return ResponseEntity.ok(Map.of("message", "Stock confirmed successfully"));
    }
}

File: backend/inventory-service/src/controllers/inventoryController.js
const inventoryService = require('../services/inventoryService');

class InventoryController {
  // Create product
  async createProduct(req, res) {
    try {
      const product = await inventoryService.createProduct(req.body);
      res.status(201).json(product);
    } catch (error) {
      res.status(400).json({ message: error.message });
    }
  }

  // Get product by ID
  async getProductById(req, res) {
    try {
      const product = await inventoryService.getProductById(req.params.id);
      res.status(200).json(product);
    } catch (error) {
      res.status(404).json({ message: error.message });
    }
  }

  // Get all products or by category
  async getAllProducts(req, res) {
    try {
      const { category } = req.query;
      const products = category
        ? await inventoryService.getProductsByCategory(category)
        : await inventoryService.getAllProducts();
      res.status(200).json(products);
    } catch (error) {
      res.status(500).json({ message: error.message });
    }
  }

  // Update product
  async updateProduct(req, res) {
    try {
      const product = await inventoryService.updateProduct(req.params.id, req.body);
      res.status(200).json(product);
    } catch (error) {
      res.status(400).json({ message: error.message });
    }
  }

  // Delete product
  async deleteProduct(req, res) {
    try {
      const result = await inventoryService.deleteProduct(req.params.id);
      res.status(200).json(result);
    } catch (error) {
      res.status(404).json({ message: error.message });
    }
  }

  // Check stock (inter-service endpoint)
  async checkStock(req, res) {
    try {
      const { productId, quantity } = req.body;
      const result = await inventoryService.checkStock(productId, quantity);
      res.status(200).json(result);
    } catch (error) {
      res.status(400).json({ message: error.message });
    }
  }

  // Reserve stock (inter-service endpoint)
  async reserveStock(req, res) {
    try {
      const { productId, quantity } = req.body;
      const result = await inventoryService.reserveStock(productId, quantity);
      res.status(200).json(result);
    } catch (error) {
      res.status(400).json({ message: error.message });
    }
  }

  // Confirm stock (inter-service endpoint)
  async confirmStock(req, res) {
    try {
      const { productId, quantity } = req.body;
      const result = await inventoryService.confirmStock(productId, quantity);
      res.status(200).json(result);
    } catch (error) {
      res.status(400).json({ message: error.message });
    }
  }

  // Release stock (inter-service endpoint)
  async releaseStock(req, res) {
    try {
      const { productId, quantity } = req.body;
      const result = await inventoryService.releaseStock(productId, quantity);
      res.status(200).json(result);
    } catch (error) {
      res.status(400).json({ message: error.message });
    }
  }

  // Upload product image
  async uploadProductImage(req, res) {
    try {
      const { id } = req.params;
      
      if (!req.file) {
        return res.status(400).json({ message: 'No image file provided' });
      }

      const result = await inventoryService.uploadProductImage(id, req.file);
      res.status(200).json(result);
    } catch (error) {
      res.status(400).json({ message: error.message });
    }
  }

  // Delete product image
  async deleteProductImage(req, res) {
    try {
      const { id } = req.params;
      const result = await inventoryService.deleteProductImage(id);
      res.status(200).json(result);
    } catch (error) {
      res.status(400).json({ message: error.message });
    }
  }

  // Get low stock products
  async getLowStockProducts(req, res) {
    try {
      const threshold = parseInt(req.query.threshold) || 10;
      const products = await inventoryService.getLowStockProducts(threshold);
      res.status(200).json(products);
    } catch (error) {
      res.status(500).json({ message: error.message });
    }
  }
}

module.exports = new InventoryController();
